### DISCUSSION TITLE
crew performance on crew_controller_local() with cmdstan

### DISCUSSION THREAD

**Discussion by yonisidi** [wlandau/crew#182] [General]
I am not sure which bucket this falls into...

I have a joint model in stan (brms) that takes ~400s to warmup and sample in an interactive session.

I am running a kfold (k=10) with {rsample} to evaluate posterior predictions.
  - ie fit model on training set and then predicting w the test set and evaluating with a log loss metric
  - I am pre-generating the model `so` object to avoid recompiling on each fold

If I try to run it with crew for 1 of the folds in a child session it retains the performance (~400s)

```r
controller$push(name = "fit", command = {
  library(brms)
  library(cmdstanr)
  library(targets)
  library(tidymodels)
  tidymodels::tidymodels_prefer()
  conflicted::conflicts_prefer(dplyr::lag)
  
  options(cmdstanr_write_stan_file_dir = here::here('script/pipelines/Model/stan'))
  options(mc.cores = 4)
  
  update(brms_object,
         iter = 2000, 
         thin = 2,
         cores = 4,
         seed = 12345,
         init = runif(1, 0, 1e-2),
         newdata = model_cv_prepped$recipes[[1]] |>
           recipes::bake(new_data = NULL))
})
```

But when I try to use a controller in targets that fits 3 folds at a time

```r
controller_stan <- crew::crew_controller_local(
  name = 'stan',
  workers = 3,
  seconds_idle = 60,
  local_log_directory = 'script/pipelines/crew/log',
  local_log_join = FALSE
)
```

performance drops to ~30min warmup+fit, where most of the drop is in the warmup.

I have tried to do more brms pre-processing and am extracting the diagonal mass matrix from a single chain warmup only run to pass in the inv_metric on the sampling step.

When I look at the stderr/stdout it looks like everything is running ok in the child sessions

<details>
<summary>what that looks like</summary>

```r
options(cmdstanr_write_stan_file_dir = here::here('script/pipelines/Model/stan'))

fit_obj <- update(brms_object,
                   warmup = 1000,
                   iter = 1000,
                   chains = 1,
                   save_warmup =TRUE,
                   seed = 12345,
                   init = runif(1, 0, 1e-2),
                   sig_figs = 18,
                   newdata = model_cv_prepped$recipes[[1]] |>
                     recipes::bake(new_data = NULL) )
            
brms_inv_mat <- (fit_obj[["fit"]] |> 
              rstan::get_adaptation_info() |> 
              (\(x) gsub('\\# ','',x))() |> 
              strsplit('\n'))[[1]][3] |> 
            (\(x) sprintf('c(%s)',x))() |> 
            (\(x) parse(text = x))() |> 
            eval() |>
            as.vector()

update(brms_object,
                iter = 2000, 
                thin = 2,
                cores = 4,
                seed = 12345,
                init = stats::runif(1, 0, 1e-2),
                inv_metric = brms_inv_mat,
                newdata = model_cv_prepped$recipes[[1]] |>
                  bake(new_data = NULL)
              )
```

</details>

In the interactive session run that drops my sampling times to ~200s, but I am seeing a similar loss of efficiency in the targets run ~15min.

I am still trying to find a way to reprex this in a canonical way but wanted to ask in the meantime if you have seen something like this previously and know if this is just resources on the instance maxing out and stan can't be run in parallel locally like I am pushing my machine at the moment.

**Comment by yonisidi** [comment]
another way to see it is through average fit times as a fn of workers


```r
tibble::tibble(
  time = c(cumsum(rep(200, 10)), cumsum(rep(515, 5)), cumsum(c(rep(1000, 3), 200))),
  fit = rep(c(1,2,3,1), times = c(10, 5, 3,1)),
  workers = rep(c(1,2,3), times = c(10, 5, 4))
  ) |>
  dplyr::mutate(
    i = cumsum(fit),
    .by = workers
  ) |>
  ggplot(aes(x = i, y = time, colour = factor(workers))) +
  geom_point() + 
  geom_line() +
  scale_x_continuous(breaks = seq(10)) +
  labs(
    x = 'Total Fits',
    y = 'Cumulative Time',
    colour = 'Parallel Workers'
  ) +
  theme(
    legend.position = 'top'
  )
```

![image](https://github.com/user-attachments/assets/7e5fa9ea-2e13-42df-9808-a0fce64032ed)

**Reply by wlandau** [reply]
How many cores does your local machine have? If Stan is already running with `cores = 4`, then 3 `crew` workers may mean 12 total Stan processes are competing for however many cores are on your laptop. Even with hyperthreading, most people experience marginal gains after exceeding `n_cores` total processes.

**Reply by yonisidi** [reply]
machine has 16 cores and it looks like the 12 are being engaged properly based on `top` and the run logs

But as you see from the plot output even with 2 parallel runs efficiency is below compared to a sequential performance. 

```
> parallel::detectCores()
[1] 16
```

```
top - 08:22:08 up 14 min,  0 users,  load average: 0.59, 1.18, 0.81
Tasks:  24 total,  14 running,  10 sleeping,   0 stopped,   0 zombie
%Cpu(s): 25.3 us,  0.1 sy,  0.0 ni, 74.6 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 65021440 total, 47705244 free,  5284764 used, 12031432 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 59012200 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                     
   1201 u1037187  20   0   31020  13684   4728 R  33.6  0.0   0:11.63 model_844588d86             
   1203 u1037187  20   0   31028  13856   4672 R  33.2  0.0   0:11.26 model_07bd53c30             
   1209 u1037187  20   0   31084  13868   4756 R  33.2  0.0   0:11.33 model_07bd53c30             
   1199 u1037187  20   0   31020  13692   4736 R  32.9  0.0   0:11.55 model_844588d86             
   1198 u1037187  20   0   31044  13784   4744 R  31.9  0.0   0:11.76 model_844588d86             
   1204 u1037187  20   0   31076  13616   4684 R  31.9  0.0   0:11.29 model_07bd53c30             
   1207 u1037187  20   0   31028  13728   4804 R  31.9  0.0   0:11.18 model_07bd53c30             
   1211 u1037187  20   0   31084  13844   4724 R  31.9  0.0   0:11.02 model_07bd53c30             
   1213 u1037187  20   0   31084  13876   4760 R  31.9  0.0   0:11.02 model_07bd53c30             
   1205 u1037187  20   0   31028  13900   4712 R  31.6  0.0   0:11.50 model_07bd53c30             
   1210 u1037187  20   0   31084  13736   4616 R  31.2  0.0   0:11.13 model_07bd53c30             
   1197 u1037187  20   0   31020  13768   4736 R  30.9  0.0   0:11.63 model_844588d86             
    406 u1037187  20   0 5855708 900872  47680 S   8.0  1.4   0:25.41 rsession                    
    964 u1037187  20   0 5148344 585144  28284 S   2.0  0.9   0:13.89 R                           
    966 u1037187  20   0 5149100 585548  28428 R   2.0  0.9   0:13.83 R                           
    975 u1037187  20   0 5148756 585088  28280 S   2.0  0.9   0:13.87 R                           
    197 u1037187  20   0  298508  14840  11644 S   0.7  0.0   0:02.30 rserver                     
    902 u1037187  20   0 3953792 221820   9008 S   0.3  0.3   0:01.73 R                           
      1 u1037187  20   0     204      4      0 S   0.0  0.0   0:00.00 s6-svscan                   
     35 u1037187  20   0     204      4      0 S   0.0  0.0   0:00.00 s6-supervise                
    194 u1037187  20   0     204      4      0 S   0.0  0.0   0:00.00 s6-supervise                
    405 u1037187  20   0   11236   2668   2452 S   0.0  0.0   0:00.01 bash                        
    461 u1037187  20   0   13624   3768   3140 S   0.0  0.0   0:00.03 bash                        
    731 u1037187  20   0   57768   4164   3628 R   0.0  0.0   0:00.08 top                  
```

```

R version 4.2.1 (2022-06-23) -- "Funny-Looking Kid"
Copyright (C) 2022 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

- Project '~/trials' loaded. [renv 1.0.7]
> crew::crew_worker(settings = list(url = "ws://127.0.0.1:46091/2/4a31c7bb415292a653bd4b2a", autoexit = 15L, cleanup = 1L, output = TRUE, maxtasks = Inf, idletime = 60000, walltime = Inf, timerstart = 0L, tls = NULL, rs = c(10407L, 493332475L, 1346206533L, -360589285L, -1335015782L, 1170645562L, 1124262649L)), launcher = "stan", worker = 2L, instance = "4a31c7bb415292a653bd4b2a")
[conflicted] Will prefer dplyr::lag over any other package.
Running MCMC with 4 parallel chains...

Chain 1 Iteration:    1 / 2000 [  0%]  (Warmup) 
Chain 2 Iteration:    1 / 2000 [  0%]  (Warmup) 
Chain 3 Iteration:    1 / 2000 [  0%]  (Warmup) 
Chain 4 Iteration:    1 / 2000 [  0%]  (Warmup) 
Chain 3 Iteration:  100 / 2000 [  5%]  (Warmup) 
Chain 4 Iteration:  100 / 2000 [  5%]  (Warmup) 
Chain 1 Iteration:  100 / 2000 [  5%]  (Warmup) 
Chain 2 Iteration:  100 / 2000 [  5%]  (Warmup) 
Chain 3 Iteration:  200 / 2000 [ 10%]  (Warmup) 
Chain 4 Iteration:  200 / 2000 [ 10%]  (Warmup) 
...
```

**Reply by wlandau** [reply]
Since you're running this locally, how does performance compare to using `mirai` without `crew`, or `parallel::mclapply()`?

```r
library(mirai)
daemons(n = 3)
task1 <- mirai({...})
task2 <- mirai({...})
task3 <- mirai({...})
```

**Reply by yonisidi** [reply]
after asking more q's to devops it seems like there was an unwritten constraint on users that caps the number of cores a user can use on the shared k8s instances, which explains why I was getting these results.

**Reply by wlandau** [reply]
Ah, glad you figured it out.
